services/shhext/dedup/cache.go

```diff
@@ -0,0 +1,111 @@
+package dedup
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto/sha3"
+	whisper "github.com/ethereum/go-ethereum/whisper/whisperv6"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/util"
+)
+
+// cache represents a cache of whisper messages with a limit of 2 days.
+// the limit is counted from the time when the message was added to the cache.
+type cache struct {
+	db  *leveldb.DB
+	now func() time.Time
+}
+
+func newCache(db *leveldb.DB) *cache {
+	return &cache{db, time.Now}
+}
+
+func (d *cache) Has(filterID string, message *whisper.Message) (bool, error) {
+	has, err := d.db.Has(d.keyToday(filterID, message), nil)
```

Comment:
so you read everytime from fs when filter rpc call is executed? leveldb must be able to hold certain amount of data in memory, is that behaviour enabled by default?

Reply:
probably.
Right now the deduplication process takes about 24 ms (in that benchmark in tests), that includes both reading and writing.

Reply:
i think it must keep some data that is used for reads in-memory then, number is very low